### ISSUE TITLE
animation caching crashes visit

### ISSUE THREAD

**Issue by charlsonkim** [visit-dav/live-customer-response#285]
the animation caching crashes visit after a few hundred frames.  no obvious error file is produce, only a message at the end to contact developers

**Comment by brugger1** [comment]
The animation caching tends to use a lot of memory since it stores the geometry in memory for all the time states. Because of this it can run out of memory and crash. As it makes the first pass through the time states, the memory will increase as it caches more geometry. Once it has made its first pass through the time states the memory usage should remain fixed.

When VisIt crashed did the compute engine only crash or did all of VisIt crash?
Did you make it through an initial pass of all the time states?
How big is your dataset (number of cells or nodes) and what type of plots did you make?

You could run VisIt with "-debug 5" on the command line to get an idea of the amount of memory usage in the engine. After it has crashed you can do a "grep rss A.engine_par.*.5.vlog" and you should see the progression of memory usage. We don't have anything in the viewer that shows memory usage, but you could monitor the memory usage in the viewer by pausing the playing every once in a while and looking at the memory usage with "ps -u".

**Comment by charlsonkim** [comment]
OK.
all of visit crashes.
no, i only made it through ~1/5 of the time states.
individual datasets are ~40M , ~900 time states

200 time states* 40M ~ 8G  , i have 16G RAM  but only 2G of swap.  would increasing swap help?
![movie0200](https://user-images.githubusercontent.com/11322613/122840137-61176e00-d2ae-11eb-9c2a-df293eebcaef.jpg)

**Comment by charlsonkim** [comment]
image is from a higher resolution dataset

**Comment by wexner** [comment]
I had the same problem, especially with >10 timesteps.
My workaround is to let VisIt be called by a python script and save the plot for each timestep individually. Then I combine the individual pictures to a video later.

**Comment by brugger1** [comment]
Memory usage is definitely the issue. In addition to the memory required for just the raw data, there is memory usage for intermediate results. Also since you have several plots that will also increase the memory usage.

Increasing the swap space would help but I think it would probably defeat the purpose of the animation caching. Using swap space would slow down the access to new frames, since at some point it would always need to swap in the geometry for the next image, which isn't as fast as just reading the data if it is already in memory. Also, I notice that you are using transparency, so that would increase the rendering time as well. The upshot of all of this is that the movie probably wouldn't play very fast. The advantage of having a using animation caching is that you can play the animation back within VisIt while allowing you to change the view (rotating, panning and zooming). A downside is that you would need to regenerate the movie every time you want to play it since we don't have a way to store the information and then just read it back.

I would suggest saving a movie either as an mpeg movie or saving a sequence of PNGs and then encoding those into a movie using your favorite movie encoder. The default movie encoder that comes with VisIt is an MPEG-1 encoder which doesn't give great results. If you install ffmpeg on your system then VisIt can use that for the encoding.

**Comment by brugger1** [comment]
Sorry about your frustrations. Here is just a little more information about how VisIt operates.

VisIt is made up of several components. When either the viewer or graphical user interface (GUI) crash, then your entire VisIt session dies. When the compute engine dies, VisIt is able to start a new one and recover seamlessly (other than a possible wait for compute resources if you are running them on a cluster shared with others). The viewer and GUI rarely crash, with one notable exception that you are encountering, which is running out of memory when animation caching is enabled.

**Comment by charlsonkim** [comment]
so this is a bug and i can expect it to be addressed on the next release?

**Comment by brugger1** [comment]
If your system doesn't have enough memory for the viewer to hold all the data there isn't much that can be done other than to detect when we are about to exceed memory and then pop up an error message and abandon the animation caching. Is this a solution you would find useful and do you find animation caching useful?

**Comment by charlsonkim** [comment]
yes please.  a pop-up warning would be much preferred over a crash.
yes, i do find the animation caching useful for interaction with the viz.
sometimes i want to go back and see other angles.

**Comment by brugger1** [comment]
I entered a couple of issues to VisIt issue tracker related to this.

https://github.com/visit-dav/visit/issues/5806
https://github.com/visit-dav/visit/issues/5809
